90zfsbootmenu/zfsbootmenu-lib.sh

```diff
@@ -901,6 +900,162 @@ clone_snapshot() {
   return 0
 }
 
+# arg1: filesystem name
+# arg2: snapshot name
+# prints: nothing
+# returns: 0 on success
+
+create_snapshot() {
+  local selected target pool
+
+  selected="${1}"
+  if [ -z "$selected" ]; then
+    zerror "selected is undefined"
+    return 1
+  fi
+  zdebug "selected: ${selected}"
+
+  target="${2}"
+  if [ -z "$target" ]; then
+    zerror "target is undefined"
+    return 1
+  fi
+  zdebug "target: ${target}"
+
+  pool="${selected%%/*}"
+  if ! set_rw_pool "${pool}" ; then
+    zerror "unable to set pool ${pool} read/write"
+    return 1
+  fi
+
+  load_key "${selected}"
+
+  zdebug "creating snapshot ${selected}@${target}"
+  if ! output="$( zfs snapshot "${selected}@${target}" )" ; then
+    zdebug "unable to create snapshot: ${output}"
+    return 1
+  fi
+
+  return 0
+}
+
+# arg1: selected snapshot
+# arg2: subkey
+# prints: snapshot/filesystem creation prompt
+# returns: nothing
+
+snapshot_dispatcher() {
+  local selected subkey
+  local parent_ds avail_space_exact be_size_exact leftover_space avail_space be_size
+  local prompt header check_base pre_populated user_input valid_name clone_target
+
+  selected="${1}"
+  if [ -z "$selected" ]; then
+    zerror "selected is undefined"
+    return 1
+  fi
+  zdebug "selected: ${selected}"
+
+  subkey="${2}"
+  if [ -z "$subkey" ]; then
+    zerror "subkey is undefined"
+    return 1
+  fi
+  zdebug "subkey: ${subkey}"
+
+  local parent_ds
+  parent_ds="${selected%/*}"
```

Comment:
1. I don't think we ever considered what happens if the BE is on the top-level filesystem. It seems like everything else would still work, but this probably requires
    ```sh
    parent_ds="${selected%/*}"
    parent_ds="${selected%@*}"
    ```
    to strip the snapshot name in that case.
2. Should this also test for empty `${parent_ds}` and fail in that case?

Reply:
Do you mean the BE would be at `zroot` or `zroot/void` ?

Reply:
If the BE is at `zroot`, the selected snapshot will be `zroot@snap`. If `parent_ds` tries to strip out just `/*`, it will still be `zroot@snap`, and everything that follows will probably fail. If we also strip out `@*` from `parent_ds` for the case when there is no `/` in the selected name, I think all of the snapshot and clone logic still works. It just means that clones and send/recv duplicats will be *children* of the BE (at `zroot/clone`, for example), but that should still be OK.

This is probably not a common configuration, and most definitely isn't advisable, but we can probably support it for free.

Reply:
Ah yeah. It's such an obscure/ill-advised combination that I wanted to make sure I was on the same page as you. I've added in the removal of `@snapshot`, and then a check to make sure that `parent_ds` is not empty.